### PR TITLE
Fixed assertEqualsDeep for large set comparison

### LINKED ISSUE #1942: assertDeepEquals takes long time to evaluate size mismatch
### Expected behavior
Let say we have 100 unique items inside the expected set SetA, We have 101 unique items inside actual set SetB.
This is a clear case of one extra element in the expected is not present in the actual.

Example:  
> assertEqualsDeep( SetB, SetA, "same same" );

Instead, if we compare the size of expected and actual if the expected is more than actual we can throw out the error immediately

*java.lang.AssertionError: Sets not equal*

### Actual behavior
Currently what the assertEqualsDeep method does is comparing the 100 available items in the expected that finds the 101th item is not present in the actual and throws error.

*java.lang.AssertionError: Sets not equal*

### Is the issue reproductible on runner?

- [ ] Shell
- [ ] Maven
- [x] Gradle
- [ ] Ant
- [ ] Eclipse
- [x] IntelliJ
- [ ] NetBeans

### PR DESCRIPTION
**Pull request by reachbharathan**
Fixes #1942  .

### Did you remember to?

- [x] Add test case(s)
- [x] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

**Comment by krmahadevan** [comment]
Can you please remove the wild card imports in the test ? TestNG has a code convention that refrains from using wild card imports.

File: src/main/java/org/testng/Assert.java [hunk 1]
```diff
@@ -1280,13 +1280,18 @@ public static void assertEqualsDeep(Set<?> actual, Set<?> expected, String messa
       }
     }
 
+    if(expected.size()>actual.size()){
```
Comment (krmahadevan) [inline src/main/java/org/testng/Assert.java]: A couple of things on this 

1. Please format this code snippet.
2. This condition doesn't handle the scenario wherein `expected.size() < actual.size()`. So wouldn't we be better off by just checking `expected.size() != actual.size()` ?
3. Add a new test for the less than condition as well.

**Review by krmahadevan — CHANGES_REQUESTED** [review]

File: src/main/java/org/testng/Assert.java (hunk 1 above)
Comment (reachbharathan) [reply src/main/java/org/testng/Assert.java]: #####Please find my response below @krmahadevan 
Let me know your thoughts.
1. Format completed
2. The original implementation of `assertEqualsDeep` is to compare a set of items between actual and expected, this is similar to stringContains for set, for more deatils refer 'https://github.com/cbeust/testng/issues/1917#issuecomment-435259490'
3. Added test for less than condition as well `arrayDeepInSetsAssertEquals`

Comment (krmahadevan) [reply src/main/java/org/testng/Assert.java]: > The original implementation of assertEqualsDeep is to compare a set of items between actual and expected, this is similar to stringContains for set, if we are using expected.size() != actual.size() then that is similar to assertEquals where the actual should always be exactly to expected not more not less, for more details refer '#1917 (comment)'

I dont think that's the case. `assertEquals` merely relies on the `.equals()` method and is agnostic of the underlying type of the object. IMO size check has to be done (`expected.size() != actual.size()`) to ensure there's symmetry in the comparison. Two sets should fail deep equality if they dont match in size. Only if the size matches is when we should proceed further with checking the contents of the set.

Please see https://github.com/cbeust/testng/issues/1917#issuecomment-435562562

Also the tests dont check for the error message but merely check for the assertion failure (which is going to happen anyway). Can we please refactor both the tests (`expected > actual` and `expected < actual`) that specifically check for the error message as well ?

@juherr please let me know your thoughts

In a nutshell, I believe that deepEquals

File: src/test/java/test/asserttests/ArrayEqualityAssertTest.java [hunk 2]
```diff
@@ -191,6 +191,28 @@ public void arrayInsideSetAssertEqualsDeep() {
     assertEqualsDeep(set, setCopy, "arrays inside sets are compared by value in assertEqualsDeep");
   }
 
+  @Test
+  public void arrayDeepInSetsAssertEquals() {
```
Comment (juherr) [inline src/test/java/test/asserttests/ArrayEqualityAssertTest.java]: It looks like this test is wrong because the 2 collections are not equal. What do you think?

File: src/main/java/org/testng/Assert.java (hunk 1 above)
Comment (juherr) [reply src/main/java/org/testng/Assert.java]: I agree. The 2 collections are not equals and one new test looks wrong.

**Comment by reachbharathan** [comment]
@krmahadevan @juherr Addressed the feedback, pls review.

**Comment by krmahadevan** [comment]
@reachbharathan - One of the comments that I had added was a request for adding assertion on the error message. That seems to be missing here. Everything else looks good.

@juherr - Do you think that we can live without the assertion on the error message ?

**Review by juherr — APPROVED** [review]

**Comment by juherr** [comment]
> @juherr - Do you think that we can live without the assertion on the error message ?

Yes :)

**Review by krmahadevan — APPROVED** [review]
